routes/groups.js

```diff
@@ -1,27 +1,120 @@
 const { Router, query } = require("express");
 const router = Router();
+const { isAuthorized } = require("./middleware.js")
 
-// Authentication middleware
+const groupDAO = require('../daos/groups');
 
-// Routes
-router.post("/", async (req, res, next) => {
-    // to complete
-});
+router.post("/",
+    isAuthorized, 
+    async (req, res, next) => {
+        const { name, origin, destination, isDefault } = req.body;
+        const newGroup = await groupDAO.create(req.userId, name, origin, destination, isDefault);
+        if (newGroup) {
+            res.json(newGroup);
+        } else {
+            res.sendStatus(404);
+        }
+    }
+);
 
-router.get("/", async (req, res, next) => {
-    // to complete
-});
+router.get("/", 
+    isAuthorized, 
+    async (req, res, next) => {
+        let { origin, destination } = req.query;
+        if (origin && !destination) {
+            const groups = await groupDAO.getByOrigin(origin);
+            if (groups) {
+                res.json(groups);
+            } else {
+                res.sendStatus(404);
+            }
+        } else if (destination && !origin) {
+            const groups = await groupDAO.getByDestination(destination);
+            if (groups) {
+                res.json(groups);
+            } else {
+                res.sendStatus(404);
+            }
+        } else if (origin && destination) {
+            const searchString = `${origin} ${destination}`;
+            const groups = await groupDAO.getByOriginAndDestination(searchString);
+            if (groups) {
+                res.json(groups);
+            } else {
+                res.sendStatus(404);
+            }
+        } else {
+            const groups = await groupDAO.getByUserId(req.userId);
+            if (groups) {
+                res.json(groups);
+            } else {
+                res.sendStatus(404);
+            }
+        }
+    }
+);
 
-router.get("/:id", async (req, res, next) => {
-    // to complete
-});
+router.get("/:id",
+    isAuthorized,
+    async (req, res, next) => {
+        const groupId = req.params.id;
+        if (groupId) {
+            const storedUserId = await groupDAO.getUserIdFromGroupId(groupId);
```

Comment:
Let's pass the `userId` from the request to a `getByIdAndUserId` method instead and remove one query.

Reply:
Done :)